### PR TITLE
Support Saving Tensors in Graph Mode with add_for_mode

### PR DESCRIPTION
**Pull request by NihalHarish**
### Description of changes:
- This allows users to directly add tensors with `add_for_mode` in `v1.compat` mode

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: smdebug/tensorflow/keras.py [hunk 1]
```diff
@@ -341,6 +341,14 @@ def _prepare_layers(self, mode):
             for w in weights:
                 self._check_and_add_layer_tensor(mode, layer, "weight", w)
 
+    def _prepare_non_layer_tensors(self):
+        for coll in self.collection_manager.get_collections().values():
+            for tensor_ref in coll.get_tensors():
+                if tensor_ref.name not in self.tensor_to_collections:
+                    self.tensor_to_collections[tensor_ref.name] = {coll}
+                elif coll not in self.tensor_to_collections[tensor_ref.name]:
+                    self.tensor_to_collections[tensor_ref.name].add(coll)
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:350]: if this is a set, you dont need the``` if coll not in self.tensor_to_collections```.

File: smdebug/tensorflow/keras.py [hunk 2]
```diff
@@ -341,6 +341,14 @@ def _prepare_layers(self, mode):
             for w in weights:
                 self._check_and_add_layer_tensor(mode, layer, "weight", w)
 
+    def _prepare_non_layer_tensors(self):
+        for coll in self.collection_manager.get_collections().values():
+            for tensor_ref in coll.get_tensors():
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:346]: can you call get_tensors() and save the result before the for loop?

File: tests/tensorflow2/test_functional_model.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    x = Dropout(0.5)(x)
+    out = Dense(num_classes, activation="softmax")(x)
+
+    model = tf.keras.models.Model(inputs=img_inputs, outputs=out)
+
+    smd_callback = smd.KerasHook(
+        export_tensorboard=False, out_dir=out_dir, include_collections=["custom"]
+    )
+
+    smd_callback.get_collection("custom").add_for_mode([x1], mode=smd.modes.TRAIN)
```
Comment (vandanavk) [inline tests/tensorflow2/test_functional_model.py:53]: what is the behavior if this is called with gradtape?

Comment (NihalHarish) [reply tests/tensorflow2/test_functional_model.py:53]: Gradtape is only supported inside of tf.function or when eager execution is enabled.
This method of saving tensors is only supported when in tf v1 mode.

**Review by vandanavk — APPROVED** [review]

**Pull request merged**